File: src/main/kotlin/dev/jefy/connectpro/portfolio/applicaion/query/ProjectQueryImpl.java
package dev.jefy.connectpro.portfolio.applicaion.query;


import org.jspecify.annotations.NullMarked;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.util.Assert;

import java.util.List;

import dev.jefy.connectpro.portfolio.applicaion.dtos.ProjectResponse;
import dev.jefy.connectpro.portfolio.domain.repository.ProjectRepository;
import dev.jefy.connectpro.portfolio.domain.vo.PortfolioId;
import dev.jefy.connectpro.shared.application.exceptions.ResourceNotFound;
import lombok.RequiredArgsConstructor;

/**
 * @author Jôph Yamba
 */
@NullMarked
@Service
@Transactional(readOnly = true)
@RequiredArgsConstructor
public class ProjectQueryImpl implements ProjectQuery {
    private final ProjectRepository projectRepo;
    @Override
    public ProjectResponse get(ProjectId projectId) {
        Assert.notNull(projectId, "projectId cannot be null");
        return projectRepo
                .findById(projectId)
                .map(ProjectResponse::fromDomain)
                .orElseThrow(()-> new ResourceNotFound("Project not found"));
    }

    @Override
    public List<ProjectResponse> getAllForPortfolio(PortfolioId portfolioId) {
        Assert.notNull(portfolioId, "id cannot be null");
        return projectRepo
                .findAllByPortfolioId(portfolioId)
                .stream()
                .map(ProjectResponse::fromDomain)
                .toList();
    }
}

File: src/main/kotlin/dev/jefy/connectpro/portfolio/application/query/ProjectQueryImpl.kt
package dev.jefy.connectpro.portfolio.application.query

import dev.jefy.connectpro.portfolio.application.dtos.ProjectResponse
import dev.jefy.connectpro.portfolio.application.exceptions.ProjectNotFoundException
import dev.jefy.connectpro.portfolio.domain.repository.ProjectRepository
import dev.jefy.connectpro.portfolio.domain.vo.PortfolioId
import dev.jefy.connectpro.portfolio.domain.vo.ProjectId
import dev.jefy.connectpro.shared.infrastructure.annotations.QueryService
import dev.jefy.connectpro.shared.infrastructure.file_storage.ImageUrlResolver

@QueryService
class ProjectQueryImpl(
    private val projectRepo: ProjectRepository,
    private val resolver: ImageUrlResolver
) : ProjectQuery {

    override fun get(projectId: ProjectId): ProjectResponse = projectRepo.findById(projectId)
            .map { ProjectResponse.fromDomain(it, resolver) }
            .orElseThrow { ProjectNotFoundException() }

    override fun getAllForPortfolio(portfolioId: PortfolioId): List<ProjectResponse> = projectRepo
        .findAllByPortfolioId(portfolioId)
        .map { ProjectResponse.fromDomain(it, resolver) }
    
}